nuxeo-retention-core/src/main/java/org/nuxeo/ecm/retention/service/RetentionComponent.java

```diff
@@ -0,0 +1,106 @@
+import org.nuxeo.ecm.retention.work.RetentionRecordUpdateWork;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.model.DefaultComponent;
+
+public class RetentionComponent extends DefaultComponent implements RetentionService {
+
+    public static Log log = LogFactory.getLog(RetentionComponent.class);
+
+    @Override
+    public void attachRule(String ruleId, DocumentModel doc, CoreSession session) {
+        // ToDo: check if we need the unrestricted in the service
+        // ToDo: compute min_cutoff_at and max_retention_at
+        CoreInstance.doPrivileged(session, (CoreSession s) -> {
+            if (!doc.hasFacet(RECORD_FACET)) {
+                doc.addFacet(RECORD_FACET);
+            }
+            s.saveDocument(doc);
+        });
+    }
+
+    @Override
+    public void attachRule(String ruleId, String query, CoreSession session) {
+
+        WorkManager workManager = Framework.getLocalService(WorkManager.class);
+        long offset = 0;
+        List<DocumentModel> nextDocumentsToBeUpdated;
+
+        CoreQueryDocumentPageProviderDescriptor desc = new CoreQueryDocumentPageProviderDescriptor();
+        desc.setPattern(query);
+        Map<String, Serializable> props = new HashMap<String, Serializable>();
+        props.put(CoreQueryDocumentPageProvider.CORE_SESSION_PROPERTY, (Serializable) session);
+
+        @SuppressWarnings("unchecked")
+        PageProvider<DocumentModel> pp = (PageProvider<DocumentModel>) Framework.getService(PageProviderService.class)
+                                                                                .getPageProvider("", desc, null, null,
+                                                                                        batchSize, 0L, props);
+        final long maxResult = pp.getPageSize();
+        do {
+            pp.setCurrentPageOffset(offset);
+            pp.refresh();
+            nextDocumentsToBeUpdated = pp.getCurrentPage();
+            if (nextDocumentsToBeUpdated.isEmpty()) {
+                break;
+            }
+            List<String> docIds = nextDocumentsToBeUpdated.stream()
+                                                          .map(DocumentModel::getId)
+                                                          .collect(Collectors.toList());
+            // ToDo: check to see if we really need an worker depending on what we need to process on the
+            // document later
+            RetentionRecordUpdateWork work = new RetentionRecordUpdateWork(ruleId);
+            work.setDocuments(session.getRepositoryName(), docIds);
+            workManager.schedule(work, WorkManager.Scheduling.IF_NOT_SCHEDULED, true);
+            offset += maxResult;
+        } while (nextDocumentsToBeUpdated.size() == maxResult && pp.isNextPageAvailable());
+    }
+
+    final class CoreQueryDocumentPageProviderDescriptor extends GenericPageProviderDescriptor {
+        private static final long serialVersionUID = 1L;
```

Comment:
Why don't you use org.nuxeo.ecm.platform.query.core.CoreQueryPageProviderDescriptor?